plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/modelsourcewatcher/FileWatcher.java

```diff
@@ -0,0 +1,195 @@
+/********************************************************************************
+ * Copyright (c) 2020 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+package org.eclipse.glsp.server.features.modelsourcewatcher;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.glsp.server.actions.ActionDispatcher;
+import org.eclipse.glsp.server.model.GModelState;
+import org.eclipse.glsp.server.protocol.ClientSessionListener;
+import org.eclipse.glsp.server.protocol.ClientSessionManager;
+import org.eclipse.glsp.server.protocol.GLSPClient;
+import org.eclipse.glsp.server.utils.ClientOptions;
+
+import com.google.inject.Inject;
+
+public class FileWatcher implements ClientSessionListener, ModelSourceWatcher {
+
+   private static final String FILE = "file://";
+
+   @Inject
+   private ActionDispatcher actionDispatcher;
+
+   private final Map<String, List<FileWatchWorker>> workers = new HashMap<>();
+
+   @Inject
+   public FileWatcher(final ClientSessionManager sessionManager) {
+      sessionManager.addListener(this);
+   }
+
+   public FileWatcher(final ClientSessionManager sessionManager, final ActionDispatcher actionDispatcher) {
+      this(sessionManager);
+      this.actionDispatcher = actionDispatcher;
+   }
+
+   @Override
+   public void sessionClosed(final String clientId, final GLSPClient client) {
+      disposeAllWorkers(clientId);
+   }
+
+   @Override
+   public void startWatching(final GModelState modelState) {
+      createWorkers(modelState).forEach(FileWatchWorker::start);
```

Comment:
Not sure how expensive that is but would it make sense to check whether the existing workers are already watching the respective files?

Reply:
Imho this would complicate the thread management as we now create all workers on start and always start from a clean slate if invoked again. I don't think it is expensive, but I haven't measured it.